gpflow/utilities/monitor.py

```diff
@@ -0,0 +1,231 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+""" Provides basic functionality to monitor optimisation runs """
+
+
+from abc import ABC, abstractmethod
+from io import BytesIO
+from typing import Any, Callable, Dict, List, Optional, Union
+
+import numpy as np
+import tensorflow as tf
+from matplotlib.backends.backend_agg import FigureCanvasAgg
+from matplotlib.figure import Axes, Figure
+
+from ..models import BayesianModel
+from .utilities import parameter_dict
+
+
+class TasksCollection:
+    """
+    Aggregates a list of tasks and allows to run them all with one call to `__call__`.
+    """
+
+    def __init__(self, tasks: List["MonitorTask"]):
+        """
+        :param tasks: a list of `MonitorTasks`s to be ran.
+        """
+        self.tasks = tasks
+
+    def __call__(self, step, **kwargs):
+        for task in self.tasks:
+            task(step, **kwargs)
+
+
+class MonitorTask(ABC):
+    """
+    A base class for a monitoring task.
+
+    All monitoring tasks are callable objects.
+    A descendant class must implement the `run` method, which is the body of the monitoring task.
+    """
+
+    def __init__(self, period: int = 1):
+        """
+        :param period: Interval between triggering the task.
+        """
+        self.current_step = 0
+        self.period = period
+
+    def __call__(self, step: int, **kwargs):
+        """
+        It calls the 'run' function and sets the current step.
+
+        :param step: current step in the optimisation.
+        :param kwargs: additional key-word arguments that can be passed
+            to the `run` method of the task. This is in particular handy for
+            passing keyword argument to the callback of `ScalarToTensorBoardTask`.
+        """
+        if step % self.period == 0:
+            self.current_step = step
+            self.run(**kwargs)
+
+    @abstractmethod
+    def run(self, **kwargs):
+        """
+        Implements the task to be executed on __call__.
+        The current step is available through `self.current_step`.
+
+        :param kwargs: keyword arguments available to the run method.
+        """
+        raise NotImplementedError
+
+
+class TensorBoardTask(MonitorTask):
+    def __init__(self, log_dir: str, period: int = 1):
+        """
+        :param log_dir: directory in which to store the tensorboard files.
+            Can be a nested. E.g. ./logs/my_run/
+        :param period: interval at which to run the task.
+            For large values of`period` the task will be less frequently ran.
```

Comment:
```suggestion
            For large values of `period` the task will be less frequently run.
```
would it be more clear to just write
```
:param period: defines how often to run the task; it will execute every `period`th step.
```
?

Reply:
thanks. done